api/types/authentication.go

```diff
@@ -295,121 +296,263 @@ type U2F struct {
+
+func (d *MFADevice) MarshalJSON() ([]byte, error) {
+	buf := new(bytes.Buffer)
+	err := (&jsonpb.Marshaler{}).Marshal(buf, d)
+	return buf.Bytes(), trace.Wrap(err)
+}
+
+func (d *MFADevice) UnmarshalJSON(buf []byte) error {
+	return jsonpb.Unmarshal(bytes.NewReader(buf), d)
+}
+
+// NewU2FDevice creates a TOTP MFADevice from the given key.
+func NewTOTPDevice(name, key string) (*MFADevice, error) {
+	d := newMFADevice(name)
+	d.Device = &MFADevice_Totp{Totp: &TOTPDevice{
+		Key: key,
+	}}
+	if err := d.CheckAndSetDefaults(); err != nil {
 		return nil, trace.Wrap(err)
 	}
-	return &u2f.Registration{
-		Raw:       l.U2FRegistration.Raw,
-		KeyHandle: l.U2FRegistration.KeyHandle,
-		PubKey:    *pubKey,
-	}, nil
+	return d, nil
+}
+
+func (d *TOTPDevice) Check() error {
+	if d.Key == "" {
+		return trace.BadParameter("TOTPDevice missing Key field")
+	}
+	return nil
 }
 
-// SetU2FRegistration encodes and stores a u2f registration.  Use nil to
-// delete an existing registration.
-func (l *LocalAuthSecrets) SetU2FRegistration(reg *u2f.Registration) error {
-	if reg == nil {
-		l.U2FRegistration = nil
-		return nil
+func (d *TOTPDevice) Equals(other *TOTPDevice) bool {
+	if (d == nil) || (other == nil) {
+		return (d == nil) && (other == nil)
 	}
-	pubKeyDer, err := x509.MarshalPKIXPublicKey(&reg.PubKey)
+	return d.Key == other.Key
+}
+
+// NewU2FDevice creates a U2F MFADevice object from a completed U2F
+// registration.
+func NewU2FDevice(name string, reg *u2f.Registration) (*MFADevice, error) {
+	d := newMFADevice(name)
+	pubKey, err := x509.MarshalPKIXPublicKey(&reg.PubKey)
 	if err != nil {
-		return trace.Wrap(err)
+		return nil, trace.Wrap(err)
 	}
-	l.U2FRegistration = &U2FRegistrationData{
-		Raw:       reg.Raw,
+	d.Device = &MFADevice_U2F{U2F: &U2FDevice{
 		KeyHandle: reg.KeyHandle,
-		PubKey:    pubKeyDer,
-	}
-	if err := l.U2FRegistration.Check(); err != nil {
-		return trace.Wrap(err)
+		PubKey:    pubKey,
+	}}
+	if err := d.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
 	}
-	return nil
+	return d, nil
 }
 
-// Check validates local auth secret members.
-func (l *LocalAuthSecrets) Check() error {
-	if len(l.PasswordHash) > 0 {
-		if _, err := bcrypt.Cost(l.PasswordHash); err != nil {
-			return trace.BadParameter("invalid password hash")
-		}
+func (d *U2FDevice) Check() error {
+	if len(d.KeyHandle) == 0 {
+		return trace.BadParameter("U2FDevice missing KeyHandle field")
 	}
-	if len(l.TOTPKey) > 0 {
-		if _, err := totp.GenerateCode(l.TOTPKey, time.Time{}); err != nil {
-			return trace.BadParameter("invalid TOTP key")
-		}
+	if len(d.PubKey) == 0 {
+		return trace.BadParameter("U2FDevice missing PubKey field")
 	}
-	if l.U2FRegistration != nil {
-		if err := l.U2FRegistration.Check(); err != nil {
-			return trace.Wrap(err)
-		}
+	if _, err := d.GetPubKeyDecoded(); err != nil {
+		return trace.BadParameter("U2fDevice PubKey is invalid: %v", err)
```

Comment:
```suggestion
		return trace.BadParameter("U2FDevice PubKey is invalid: %v", err)
```
for consistency with other messages.